Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -26,51 +26,368 @@
+        return true;
+    
+    // TODO: Is a select element whose multiple content attribute is absent.
+    
+    return false;
 }
 
-static std::optional<SimpleRange> rangeOfString(const String& target, Document& document)
+static bool isNonSearchableSubtree(Node& node)
 {
-    if (target.isEmpty())
+    Node* traversingNode = &node;
+    while(traversingNode) {
+        if (isSearchInvisible(*traversingNode))
+            return true;
+        traversingNode = traversingNode->parentInComposedTree();
+    }
+    return false;
+}
+
+static Node& nearestBlockAncestor(Node& node)
+{
+    Node* curNode = &node;
+    while (curNode)
+    {
+        if (!curNode->isTextNode() && curNode->renderer() && curNode->renderer()->style().isDisplayBlockLevel())
+            return *curNode;
+        else
+            curNode = curNode->parentInComposedTree();
+    }
+    return node.document();
+}
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<Node*> textNodeList, bool isEnd)
+{
+    unsigned counted = 0;
+    
+    for (auto* curNode : textNodeList) {
+        unsigned nodeEnd = counted + curNode->length();
+        if (isEnd)
+            nodeEnd++;
+        if (nodeEnd > index)
+            return BoundaryPoint(*curNode, index - counted);
+        counted += curNode->length();
+    }
+    return std::nullopt;
+}
+
+static bool indexIsWordBoundary(const String& string, unsigned index)
+{
+    UBreakIterator* it = wordBreakIterator(string);
+    return ubrk_isBoundary(it, index);
+}
+
+static bool isVisibleTextNode(Node& node)
+{
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, SimpleRange searchRange, Vector<Node*> nodes, bool wordStartBounded, bool wordEndBounded)
+{
+    String searchBuffer;
+    std::optional<BoundaryPoint> start;
+    std::optional<BoundaryPoint> end;
+    
+    if (nodes.size() < 1)
         return std::nullopt;
+    
+    StringBuilder searchBufferBuilder;
+    for (auto* node : nodes) {
+        if (node)
+            searchBufferBuilder.append(downcast<Text>(*node).data());
+    }
+    searchBuffer = searchBufferBuilder.toString();
+    
+    unsigned searchStart = 0;
+    
+    if (nodes[0] == &searchRange.startContainer())
+        searchStart = searchRange.startOffset();
+    
+    std::optional<unsigned> matchIndex;
+    
+    while (!matchIndex) {
+        // TODO: find only using the base characters i.e. also fold accents and others.
+        auto potentialIndex = searchBuffer.findIgnoringASCIICase(query, searchStart);
 
-    auto searchRange = makeRangeSelectingNodeContents(document);
+        if (potentialIndex == notFound)
+            return std::nullopt;
+        else
+            matchIndex = potentialIndex;
+        
+        start = getBoundaryPointAtIndex(*matchIndex, nodes, false);
```

Comment:
We should be using enum classes to avoid obscure boolean parameters like these. No idea what `false` means by just looking at the call site.